Fix authorization issue - when request is denied return forbbiden exist code (403).

liron-l opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**
When request is rejected by authorization plugin, replace status code 500 (internal server error) with 403 (forbidden). Based on discussion in #2242

**\- How I did it**
Used `errors.NewErrorWithStatusCode` when authorization plugin rejects requests

**\- How to verify it**
New integration test for API

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**\- A picture of a cute animal (not mandatory but encouraged)**
- Fix outdated documentation on response modification in authorization
- Return 403 (forbidden) when request is denied in authorization flows
  (including integration test)
- Fix #22428

Signed-off-by: Liron Levin liron@twistlock.com

Commit:
Fix authorization issue - when request is denied return forbbiden exist code (403).

- Return 403 (forbidden) when request is denied in authorization flows
(including integration test)
- Fix #22428
- Close #22431

Signed-off-by: Liron Levin <liron@twistlock.com>

thaJeztah commented on pkg/authorization/authz.go:
```diff
@@ -85,7 +86,7 @@ func (ctx *Ctx) AuthZRequest(w http.ResponseWriter, r *http.Request) error {
 
 		if !authRes.Allow {
-			return fmt.Errorf("authorization denied by plugin %s: %s", plugin.Name(), authRes.Msg)
+			return errors.NewErrorWithStatusCode(fmt.Errorf("authorization denied by plugin %s: %s", plugin.Name(), authRes.Msg), http.StatusForbidden)
```
Perhaps create a new helper function for this? (`NewRequestForbiddenError`?) Similar to https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/errors/errors.go#L37-L41

(also two other locations may need to be updated then; https://github.com/docker/docker/blob/7fd53f7c711474791ce4292326e0b1dc7d4d6b0f/daemon/network.go#L100 and https://github.com/docker/docker/blob/7fd53f7c711474791ce4292326e0b1dc7d4d6b0f/daemon/network.go#L224)

thaJeztah commented on integration-cli/docker_cli_authz_unix_test.go:
```diff
@@ -272,6 +275,27 @@ func (s *DockerAuthzSuite) TestAuthZPluginDenyRequest(c *check.C) {
+	s.ctrl.reqRes.Allow = false
+	s.ctrl.resRes.Msg = unauthorizedMessage
+
+	daemonUrl, err := url.Parse(s.d.sock())
```
Looks like golint is complaining here; this should be `daemonURL`;

```
10:56:18 ---> Making bundle: validate-lint (in bundles/1.12.0-dev/validate-lint)
10:56:18 Errors from golint:
10:56:18 integration-cli/docker_cli_authz_unix_test.go:285:2: var daemonUrl should be daemonURL
10:56:18 
10:56:18 Please fix the above errors. You can test via "golint" and commit the result.
```

liron-l commented:
Thanks @thaJeztah, fixed both issues (also refactored network errors to use `NewRequestForbiddenError`)

thaJeztah commented:
Thanks @liron-l  - sorry, I just thought of this, but can you move the documentation changes to a separate PR? I have a cherry-pick PR to update the docs and having a separate PR makes it easier for me to cherry-pick just those changes (https://github.com/docker/docker/pull/22421) in case we do an update of the documentation before the next release

thaJeztah commented:
owww, erm.. looks like using the "errors" in a package wasn't a good idea :cry: :blush:

```
11:49:23 ---> Making bundle: validate-pkg (in bundles/1.12.0-dev/validate-pkg)
11:49:24 These files import internal code: (either directly or indirectly)
11:49:24  - pkg/authorization/authz.go imports github.com/docker/docker/errors
```

liron-l commented:
Thanks @thaJeztah, removed the reference from `pkg`, but kept the one in `daemon`.
I've also removed the documentation section and will update those in a separate PR (after this PR will pass)

thaJeztah commented:
oh looks like there's a gofmt issue now (well, at least we have managed to trigger all checks 😜)

```
12:37:25 ---> Making bundle: validate-gofmt (in bundles/1.12.0-dev/validate-gofmt)
12:37:25 These files are not properly gofmt'd:
12:37:25  - pkg/authorization/authz.go
12:37:25 
12:37:25 Please reformat the above files using "gofmt -s -w" and commit the result.
```

thaJeztah commented on pkg/authorization/authz.go:
```diff
@@ -163,3 +163,16 @@ func headers(header http.Header) map[string]string {
+	error
+}
+
+func (e authorizationError) HTTPErrorStatusCode() int {
```
Should probably be documented, as it's an exported function;

    // HTTPErrorStatusCode returns a status code.

liron-l commented:
Thanks @thaJeztah , I think everything is fixed now.

thaJeztah commented:
Yes, looks okay now. win2lin failure is not related

LGTM

lblackstone commented:
Minor nit, but the commit message still references the doc change which got pulled into a separate PR.

liron-l commented:
Thanks @lblackstone, I've updated the commit message.

aaronlehmann commented:
LGTM
